---
title: "LAAA on SARP BDC"
author: "Corey Cox"
date: "3/22/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(readxl)

`%!in%` <- Negate(`%in%`)
duplicated.all <- function(x) { duplicated(x) | duplicated(x, fromLast = TRUE) }
```

# SARP LAAA on BioDataCatalyst
We'll be performing fine mapping using LAAA on BioDataCatalyst for SARP admixture reruns for the over 12 participants in SARP1/2 + CSGA.

## Data Locations and Info
Locations for phenotype file, imputed vcf's, wgs vcf's, and original SARP local ancestry.

### Phenotype file
* `Complete SARP CSGA 3-4-21 cut.xlsx`
  - Complete "locked" phenotype file for SARP 1/2,  CSGA and SARP 3
  - Received from Victory Ortega 4-Mar-2021
  - Cleanup and select variables and re-write to `Complete_SARP_CSGA_3-4-21_cut.adjusted.20200308.tsv`

```{r cleanup_pheno, echo=FALSE}
raw_phenos <- "~/Documents/Ortega/input_admixture_mapping/Complete SARP CSGA 3-4-21 cut.xlsx" %>% read_excel()

phenos <- raw_phenos %>%
  select(`ID SARP1-2 preferred`, `analysis_set (sarp1-2/csga = 1, SARP3 = 2)`, SARP_CSGA_genetics1_FID, TopMed_ID, `Study group`,
         SITE, `Clinical Center`, `ALL COHORT Based on COMBINED Analysis`, `ASTHMA (2=yes, 1 = no)`, `RACE ( AA=1,CC=2,HS=3,Oth=4)`,
         `SARP ethnic group`, SEX, `AGE>=12`, `AGE (csga-SARP12 default. Yellow SARP3)`, `HT (csga-SARP12 default. Yellow SARP3)`,
         `BMI(csga-SARP12 default. Yellow SARP3)`, `ALL bFEV1`, `ALL bFVC`, `FEV1-FVC ratio (SARP3/yellow derived)`, `ALL MaxFEV1`,
         `ALL MaxFVC`, `ALL MaxRatio`, `ALL MaxREV...137`, `ALL pdifffev1 (SARP1-2 CSGA only)`, `ALL MaxREV...243`, `puffs to max`,
         SARP_CSGA_genetics1_FID) %>%
  mutate(group = `Study group`,
         Clinical.Center = `Clinical Center`,
         ASTHMA..2.yes..1...no. = `ASTHMA (2=yes, 1 = no)`,
         sarp12_csga_age = `AGE (csga-SARP12 default. Yellow SARP3)`,
         sarp12_csga_HTcm = `HT (csga-SARP12 default. Yellow SARP3)`,
         sarp12_csga_bmi = `BMI(csga-SARP12 default. Yellow SARP3)`,
         Percent.difference.in.FEV1.after.2.puffs.albuterol..q.17g...Spir = `ALL pdifffev1 (SARP1-2 CSGA only)`,
         sarp12_csga_bFEV1 = `ALL bFEV1`,
         sarp12_csga_bFVC = `ALL bFVC`,
         Maximum.FEV1..liters..Spirometry...maxBD = `ALL MaxFEV1`,
         Maximum.FVC..liters..Spirometry...maxBD = `ALL MaxFVC`,
         maxREVbeth = `ALL MaxREV...137`,
         puffs.to.max = `puffs to max`) %>%
  mutate(id = `ID SARP1-2 preferred`,
         center = `Clinical Center`,
         asthma = `ASTHMA (2=yes, 1 = no)`,
         height = `HT (csga-SARP12 default. Yellow SARP3)`,
         bmi = `BMI(csga-SARP12 default. Yellow SARP3)`,
         `2_puffs_albuterol_FEV1_percBD` = `ALL pdifffev1 (SARP1-2 CSGA only)`, # Ask Victor
         FEV1 = `ALL bFEV1`,
         FVC = `ALL bFVC`,
         FEV1_FVC_ratio = `FEV1-FVC ratio (SARP3/yellow derived)`,
         maxFEV1 = `ALL MaxFEV1`,
         maxFVC = `ALL MaxFVC`,
         maxFEV1_reversal = `ALL MaxREV...137`)

# phenos %>% filter(`ALL MaxREV...137` != `ALL MaxREV...243`) ## `ALL MaxREV` in two location, both are equal.

phenos%>% write_tsv("~/Documents/Ortega/input_admixture_mapping/Complete_SARP_CSGA_3-4-21_cut.adjusted.20200308.tsv")
```


### Local Ancestry and Variant Files
* BioDataCatalyst: SARP_chip_local_ancestry
  - Local ancestry already run on chr1, 8, 12
* Imputed: Pegasus- "/Volumes/Promise Pegasus/gitlab/ORTEGA/csga_imputation/data/output"
  - filenames: chr${chr}.vcf.gz
  - only samples with NO WGS data are included in imputed (so no sample overlap)
  - Use unzip_results.sh script to unzip relevant chromosomes
* WGS: Pegasus- "/Volumes/Promise Pegasus/dbgap_downloads/78082/topmed-dcc/exchange/phs001446_TOPMed_WGS_SARP/Combined_Study_Data/Genotypes/freeze.8/phased/"
  - filenames: freeze.8.chr${chr}.pass_only.phased.bcf
  - use get_sarp_chr.sh to extract relevant chromosomes
  - extract to: "/Volumes/Promise Pegasus/topmed_freeze8_phased_sarp/"
  - extract filename: chr${chr}.vcf.gz
  - rextract with current IDs

## Other notes
CSGA+SARP12 is the discovery data set and we are aiming to identify allele dose associations in the discovery data set, which we will assess replication for using the SARP3 WGS data. For BARD, we used P < 0.001 to bring forward SNPs for replication. This should be OK, but if it asked for by a reviewer, the BDC app calc_nr_indep_snps can be used to estimate the number of independent variants in a peak, and then the P-value cut-off would be 0.05/nr_indep_snps.

Chromosomes identified from SARP_admixture_rerun_no_under12:
chr1, chr8, chr9, chr11, chr16, chr19

* 2-puffs_albuterol... : chr11
* FEV1: none
* FVC: chr16
* maxFEV1: chr1
* maxFEV1_reversal: chr9, chr19
* maxFVC: chr8

```{r get_SARP_IDs}
pheno <- "~/Documents/Ortega/input_admixture_mapping/Complete_SARP_CSGA_3-4-21_cut.adjusted.20200308.tsv" %>%
  read_tsv(col_types = cols(
    .default = col_double(),
    `ID SARP1-2 preferred` = col_character(),
    TopMed_ID = col_character(),
    `Study group` = col_character(),
    SITE = col_character(),
    `Clinical Center` = col_character(),
    `ALL COHORT Based on COMBINED Analysis` = col_character(),
    `SARP ethnic group` = col_character(),
    group = col_character(),
    Clinical.Center = col_character(),
    id = col_character(),
    center = col_character()
  ))

pheno_IDs <- pheno %>% filter(! is.na(TopMed_ID) & ! duplicated(TopMed_ID) & `RACE ( AA=1,CC=2,HS=3,Oth=4)` == 1) %>% select(TopMed_ID)
pheno_IDs %>% nrow
pheno_IDs %>% write_tsv("SARP.txt", col_names = FALSE)
```

#### Checklist
- [ ] Identify which chromosomes have admixture mapping peaks for any of the phenotypes, and run RFMix for them on BDC using the  rfmix_afr_eur_imp_wf in the ORTEGA_shared_apps project (see notes above on file locations).
- [ ] Create phenotype files with outcomes and covariates needed to fine-map any significant admixture mapping peaks - create separate phenotype files for imputed data, SARP12 samples with WGS data, SARP3 samples with WGS data.

```{sh, eval=FALSE}
# Run on Pegasus and sftp to local.
# Get vcf header from imputed data for phenotype subsetting
gzcat chr19.dose.vcf.gz | head -n 20 | grep "#CHROM" > imputed_vcf_header.txt

# Get vcf header from wgs data for phenotype subsetting
gzcat chr19.vcf.gz | head -n 20 | grep "#CHROM" > wgs_vcf_header.txt
```

```{r create_phenos}
vcf_lead <- c("#CHROM", "POS", "ID", "REF", "ALT", "QUAL", "FILTER", "INFO", "FORMAT")

select_pheno <- phenos %>%
  select(TopMed_ID, FID = "SARP_CSGA_genetics1_FID", age = "sarp12_csga_age", sex = "SEX", BMI = "sarp12_csga_bmi",
    group, race = "RACE ( AA=1,CC=2,HS=3,Oth=4)", center = "Clinical.Center", height_cm = "sarp12_csga_HTcm",
    base_fev1 = "sarp12_csga_bFEV1", cohort = "ALL COHORT Based on COMBINED Analysis",
    two_puff_FEV1 = "Percent.difference.in.FEV1.after.2.puffs.albuterol..q.17g...Spir",
    FVC = "sarp12_csga_bFVC", maxFEV1 = "Maximum.FEV1..liters..Spirometry...maxBD",
    max_fev1_reversal = "maxREVbeth", maxFVC = "Maximum.FVC..liters..Spirometry...maxBD")

imputed_IDs <- "imputed_vcf_header.txt" %>% read_lines %>% str_split("\t") %>% flatten %>% unlist %>%
  magrittr::extract(. %!in% vcf_lead) %>% str_extract("^[:digit:]*")

imputed_phenos <- select_pheno %>%
  filter(FID %in% imputed_IDs, race == 1)

imputed_phenos %>% write_tsv("SARP1_2_imputed_phenos_mapping.txt")

wgs_IDs <- "wgs_vcf_header.txt" %>% read_lines %>% str_split("\t") %>% flatten %>% unlist %>%
  magrittr::extract(. %!in% vcf_lead)

wgs_phenos <- select_pheno %>% filter(TopMed_ID %in% wgs_IDs & ! duplicated(TopMed_ID))

SARP1.2_wgs_phenos <- wgs_phenos %>% filter(cohort == "SARP1-2")
SARP1.2_wgs_phenos %>% write_tsv("SARP1_2_wgs_phenos_mapping.txt")
SARP3_wgs_phenos <- wgs_phenos %>% filter(cohort == "SARP3")
SARP3_wgs_phenos %>% write_tsv("SARP3_wgs_phenos_mapping.txt")
```

- [ ] Create R model files for each outcome phenotype as documented at https://github.com/mdaya/laaa_on_sevenbridges under heading "r_model_file". Scripts Volumes/Promise Pegasus/gitlab/ORTEGA/laaa/scripts/models_*.R have the code I used to run this previously on a local machine
- [ ] On BDC, run the laaa task copied from the ORTEGA_shared_apps project for each peak to fine-map, separately for imputed data, SARP12 WGS data, SARP3 WGS data
- [ ] Combine the chip and SARP12 WGS laaa results using inverse-variance meta-analysis, for each beta, SE and P (so for allele_dose, allele_afr_dose, allele_afr_dose) - output data in a similar format as the laaa results file, so this can be used as input for locus zoom. The attached meta-analysis.R script is an analogous scenario for combining ADRN QT results for MEGA and WGS data
- [ ] On BDC, use the laaa_locus_zoom task  copied from the ORTEGA_shared_apps project to run locus zoom for CSGA+SARP12 imputed +WGS data, and SARP3 WGS data
- [ ] Identify SNPs in the CSGA+SARP12 imputed +WGS data with allele dose P < 0.001, and assess replication in SARP3 WGS data (P < 0.05, same effect direction)

