---
title: "LAAA on SARP BDC"
author: "Corey Cox"
date: "3/22/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
```

# SARP LAAA on BioDataCatalyst
We'll be performing fine mapping using LAAA on BioDataCatalyst for SARP admixture reruns for the over 12 participants in SARP1/2 + CSGA.

## Imputed data locations

* In the shared BDC project "SARP_chip_local_ancestry", local ancestry inference has already been run for chr1, 8, 12 (but depending on the new results, this may no longer be relevant)
* On Michelle's Tower: Volumes/Promise Pegasus/gitlab/ORTEGA/csga_imputation/data/output. Use the unzip_results.sh script to unzip relevant chromosomes
* Note that only samples with NO WGS data are included in the imputed data set (so no need to worry about sample overlap)

## WGS location

Use the attached get_sarp_chr.sh script - first create a file called SARP.txt that contains all the African American TOPMed sample IDs with phenotype data (each line contains a single NWD ID; I have done previous extracts but better to run it from scratch, as the list of sample IDs may be slightly different with the new phenotype file)

## Other notes

CSGA+SARP12 is the discovery data set and we are aiming to identify allele dose associations in the discovery data set, which we will assess replication for using the SARP3 WGS data. For BARD, we used P < 0.001 to bring forward SNPs for replication. This should be OK, but if it asked for by a reviewer, the BDC app calc_nr_indep_snps can be used to estimate the number of independent variants in a peak, and then the P-value cut-off would be 0.05/nr_indep_snps.

Chromosomes identified from SARP_admixture_rerun_no_under12:
chr1, chr8, chr9, chr11, chr16, chr19

* 2-puffs_albuterol... : chr11
* FEV1: none
* FVC: chr16
* maxFEV1: chr1
* maxFEV1_reversal: chr9, chr19
* maxFVC: chr8


```{r cars}
pheno <- "~/Documents/Ortega/input_admixture_mapping/Complete_SARP_CSGA_3-4-21_cut.adjusted.20200308.tsv" %>%
  read_tsv(col_types = cols(
    .default = col_double(),
    `ID SARP1-2 preferred` = col_character(),
    TopMed_ID = col_character(),
    `Study group` = col_character(),
    SITE = col_character(),
    `Clinical Center` = col_character(),
    `ALL COHORT Based on COMBINED Analysis` = col_character(),
    `SARP ethnic group` = col_character(),
    group = col_character(),
    Clinical.Center = col_character(),
    id = col_character(),
    center = col_character()
  ))

pheno %>% select(TopMed_ID) %>% write_tsv("SARP.txt", col_names = FALSE)
```

### Checklist

- [ ] >Identify which chromosomes have admixture mapping peaks for any of the phenotypes, and run RFMix for them on BDC using the  rfmix_afr_eur_imp_wf in the ORTEGA_shared_apps project (see notes above on file locations).
- [ ] Create phenotype files with outcomes and covariates needed to fine-map any significant admixture mapping peaks - create separate phenotype files for imputed data, SARP12 samples with WGS data, SARP3 samples with WGS data.
- [ ] Create R model files for each outcome phenotype as documented at https://github.com/mdaya/laaa_on_sevenbridges under heading "r_model_file". Scripts Volumes/Promise Pegasus/gitlab/ORTEGA/laaa/scripts/models_*.R have the code I used to run this previously on a local machine
- [ ] On BDC, run the laaa task copied from the ORTEGA_shared_apps project for each peak to fine-map, separately for imputed data, SARP12 WGS data, SARP3 WGS data
- [ ] Combine the chip and SARP12 WGS laaa results using inverse-variance meta-analysis, for each beta, SE and P (so for allele_dose, allele_afr_dose, allele_afr_dose) - output data in a similar format as the laaa results file, so this can be used as input for locus zoom. The attached meta-analysis.R script is an analogous scenario for combining ADRN QT results for MEGA and WGS data
- [ ] On BDC, use the laaa_locus_zoom task  copied from the ORTEGA_shared_apps project to run locus zoom for CSGA+SARP12 imputed +WGS data, and SARP3 WGS data
- [ ] Identify SNPs in the CSGA+SARP12 imputed +WGS data with allele dose P < 0.001, and assess replication in SARP3 WGS data (P < 0.05, same effect direction)


